src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -195,7 +211,35 @@ uct_cuda_copy_ctx_rsc_get(uct_cuda_copy_iface_t *iface,
         }
 
         status = uct_cuda_copy_push_ctx(cuda_device, 0, UCS_LOG_LEVEL_ERROR);
-        if (status != UCS_OK) {
+        if (ucs_unlikely(status == UCS_ERR_NO_DEVICE)) {
+            /* Device primary context of `cuda_device` is inactive. The memory
+             * was probably allocated on the context created with cuCtxCreate.
+             * Fallback to query context based on memory address. */
+            status = UCT_CUDADRV_FUNC_LOG_ERR(
+                    cuPointerGetAttribute(&cuda_context,
+                                          CU_POINTER_ATTRIBUTE_CONTEXT,
+                                          cuda_deviceptr));
+            if (status != UCS_OK) {
```

Comment:
maybe move this code to some aux function? this one seems to be quite long already.
would be good to refactor if possible
also below some comments assume that we can push only primary context which is not longer true with this PR

Reply:
Added `uct_cuda_copy_ep_push_memory_ctx`.
Updated comments.